icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t>
+  eIcicleError cpu_ntt_release_domain(const Device& device)
+  {
+    // release the memory
+    free(s_ntt_domain<S>->twiddles);
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = nullptr;
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename E = scalar_t>
+  eIcicleError bit_reverse(int size, int logn, E* output, int batch_size)
+  {
+    ICICLE_LOG_DEBUG << "BIT REVERSE";
+    int total_size = size * batch_size;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_output = output + batch * size;
+      for (int i = 0; i < size; ++i) {
+        int rev = 0;
+        for (int j = 0; j < logn; ++j) {
+          if (i & (1 << j)) { rev |= 1 << (logn - 1 - j); }
+        }
+        if (i < rev) { std::swap(current_output[i], current_output[rev]); }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+  {
+    if (size & (size - 1)) { return eIcicleError::INVALID_ARGUMENT; }
+
+    int total_size = size * config.batch_size;
+    E* temp_elements = new E[total_size];
+
+    // ICICLE_LOG_DEBUG << "INITIAL INPUT";
+    // for (int i = 0; i < total_size; ++i) {
+    //     ICICLE_LOG_DEBUG << "input[" << i << "]: " << input[i];
+    // }
+
+    if (config.columns_batch) {
+      // Distribute input into columns if columns_batch is set
+      for (int col = 0; col < config.batch_size; ++col) {
+        for (int row = 0; row < size; ++row) {
+          temp_elements[col * size + row] = input[row * config.batch_size + col];
+        }
+      }
+    } else {
+      std::copy(input, input + total_size, temp_elements);
+    }
+    // ICICLE_LOG_DEBUG << "INITIAL ELEMENTS ARRAY";
+    // for (int i = 0; i < total_size; ++i) {
+    //     ICICLE_LOG_DEBUG << "temp_elements[" << i << "]: " << temp_elements[i];
+    // }
+    const int logn = int(log2(size));
+
+    bool dit = true;
+    switch (config.ordering) { // kNN, kNR, kRN, kRR, kNM, kMN
+    case Ordering::kNN:
+      bit_reverse(size, logn, temp_elements, config.batch_size);
+      break;
+    case Ordering::kNR:
+    case Ordering::kNM:
+      dit = false; // dif
+      break;
+    case Ordering::kRR:
+      bit_reverse(size, logn, temp_elements, config.batch_size);
+      dit = false; // dif
+      break;
+    case Ordering::kRN:
+    case Ordering::kMN:
+      break;
+    default:
+      return eIcicleError::INVALID_ARGUMENT;
+    }
+
+    // ICICLE_LOG_DEBUG << "AFTER BIT REVERSE";
+    // for (int i = 0; i < total_size; ++i) {
+    //     ICICLE_LOG_DEBUG << "temp_elements[" << i << "]: " << temp_elements[i];
+    // }
+
+    S* twiddles = s_ntt_domain<S>->get_twiddles();
+
+    // NTT/INTT
+    int ntt_step = 0;
+    if (dit) {
```

Comment:
I would move dit and dif into functions to make the code easier to read

Reply:
fixed